### PR TITLE
Prometheus Integration Staging

### PR DESCRIPTION
**Pull request by connersaeli**
This PR includes staging changes for Prometheus integration. The core functionality of SUPReMM in this PR remains the same except for the addition of support for Prometheus as a data source. These changes also provide a generic template for adding support for other data sources.

### src/supremm/summarize_jobs.py
-  processjobs() now initializes a **datasource** instance from the **DatasourceFactory()** class based on the per-resource configuration.
- The instantiated **datasource** class is now a parameter for process_resource(_multiprocessing)().
- The DbAcct and XDMoDAcct classes that are instantiated here now use resconf instead of the resource_id, justification for that change explained below in account.py
- cleanjobdir() is now a concrete implementation of the **Datasource** cleanup() method for PCP. This now resides in src/supremm/datasource/pcp/pcpdatasource.py under the cleanup() method.

### src/supremm/setup.py
- Edited imports for the new directory structure which separates **datasource**-specific code
- Added **mapping.json** to configuration path

### src/supremm/Job.py
- Added nodenames() generator method. This has identical functionality to the nodearchives() generator, except that it only yields node names.
- The intent for adding this method instead of using nodearchives() is to avoid confusion with Prometheus-specific code because it only needs node names and has no concept of an archive.

### src/supremm/config.py
- Added filename parameter to autodetectconfpath() static method. The original filename lookup (config.json) is now a default for backward compatibility.
- This feature was added so the metric mapping configuration (mapping.json) can be found using this existing static method.

### src/supremm/errors.py
- Added an error for Prometheus summarization

### src/supremm/plugins/CpuUsage.py
- This plugin now checks for job-specific CPU information from both Proc and ProcPrometheus preprocessors.
- Previously only checked Proc preprocessor.

### src/supremm/proc_common.py
- Removed summarizejob() function. The main logic from this function was moved to src/supremm/datasource/pcpdatasource.py under the method name summarizejob()
- summarizejob() is now a method of the abstract **Datasource** class.

### src/supremm/supremm_setup.py
- Prompts for either pcp or prometheus as a data source when configuring different resources.
- Configuring Prometheus does a naive check to the Prometheus server with the specified settings.

### src/supremm/supremm_testharness.py
- Update import for new **datasource** directory structure
- Testharness only works for PCPSummarize

### src/supremm/account.py
- The constructor for the **Accounting** class now accepts the entire **resconf** as a parameter.
- Both **resource_id** and **hostnamemode** config settings are now used by this class so it is necessary to just pass the entire **resconf** to the constructor.

### src/supremm/xdmodaccount.py
- Constructor now accepts the entire resource configuration following changes made to **account.py**.
- Added **hostnamemode** attribute which is used to conditionally parse node names.
- Added query to get node names only. Node names are populated first then archive names are populated.

### tests/ci/setup.sh
- Added requests as a dependency to the setup script

### tests/test{summarize,PcpArchiveProcessor}.py
- Fixed imports for new directory structure

### src/supremm/indexarchives.py
- Moved to src/supremm/datasource/pcp/indexarchives.py

### src/supremm/pcparchive.py
- Moved to src/supremm/datasource/pcp/pcparchive.py

### src/supremm/pcpcinterface/
- Moved to src/supremm/datasource/pcp/

### src/supremm/summarize.py
- Renamed to pcpsummarize.py and moved to src/supremm/datasource/pcp/pcpsummarize.py
- Added some logic in get() to avoid overwriting process information from ProcPrometheus preprocessor. This bug happened because Proc and ProcPrometheus preprocessors both output the same key for the summarization record (procDump). Because PromPrometheus would always be processed after Proc, the empty output from ProcPrometheus would overwrite actual data from Proc.

### config/mapping.json
- Mapping configuration file for PCP metric names -> Prometheus metric names
- 0.0009765625 = (1 / 1024), using this decimal allows the code to just multiply all configured "scaling" strings versus having to programmatically determine which scaling factors need to be divided or multiplied.
- Previously the "scaling" configuration string (i.e. "/ 1024") was appended to the Prometheus query, but changes made to the raw data queries do not allow for any operations to be applied to API calls to Prometheus's **query** endpoint (versus the previously used **query_range**).

### src/supremm/summarize.py
Abstract summarization class that should be implemented by every data source supported by SUPReMM

**Methods:**
- process()
Main entry point. This method is called after all of the pre-summarization checks have passed.
- complete()
Returns whether or not summarization was successful for a job.
- good_enough()
Returns whether or not 95% of the nodes have been processed successfully for a job.
- get()
Returns a dictionary of the summarized job. Should be called once summarization is completed.

Concrete implementations of the **Summarize** class are free to define their own helper methods, but they must use these 4 methods to integrate into the existing SUPReMM summarization workflow.

### src/supremm/datasource/datasource.py
**Datasource** is an abstract class that provides an API for initializing summarization for individual data source backends. Each method in this class provides a generic implementation which is used by every data source. An example of this would be that every data source needs to instantiate preprocs/plugins. The concrete methods should utilize the super() function to use the generic implementation and add its own logic on top of that.

**Methods**
- presummarize()
Checks for command line options and filters jobs by these options. Returns an instance of the **JobMeta** class which is a simple container class to store metadata about the status of the job's summarization before and after summarization.
- summarizejob()
Instantiates all the preprocs and plugins. Concrete implementations should instantiate and call that **datasource**'s defined Summarize class.
- cleanup()
Implement any logic here to cleanup after summarization. For example, PCP cleans up merged archives in this step.

- summarizejob() instantiates preprocs and plugins. For each **datasource** implementation, the summarizejob() method should call super() to instantiate preproc/plugins and then use its own logic to initialize the respective **summarize** class for that **datasource**.

### src/supremm/datasource/factory.py
- Instantiates the proper concrete implementation of the **Datasource** class based on the resource's configuration. This allows summarize_jobs.py to initialize a transparent **datasource** variable and call the same methods regardless of the configured backend data source.

### src/supremm/datasource/pcp/pcpdatasource.py
- Almost all of the logic from this code was originally used in the summarizejob() function in proc_common.py and has been partitioned and slightly modified to fit in with the **Datasource** class.
- The cleanup() method was originally the cleanjobdir() function in summarize_jobs.py.

### src/supremm/datasource/prometheus/promdatasource.py
Concrete implementation of the **Datasource** class for Prometheus

**Attributes**
- client : PromClient class, handles connection to Prometheus server
- mapping : MappingManager class, handles PCP->Prometheus metric mapping configuration

**Methods**
- presummarize()
Checks to see if there summarization software can connect to the Prometheus server, else the job fails. If a connection is made, then initialize the metric mapping.
- summarizejob()
Initialize PromSummarize and check if summarization was successful.
- cleanup()
Prometheus does not require cleanup, the connection initialized in this class is persistent between summarizing jobs and the session is automatically closed by requests when done.

### src/supremm/datasource/prometheus/prominterface.py
This file contains two classes that handle retrieving data from Prometheus and reformatting those responses into a format expected by SUPReMM.

**PromClient**
Provides a simple interface to connect to Prometheus's various API endpoints. Supports configuring basic authentication to Prometheus. Other forms of authentication to Prometheus are considered experimental and are not supported at this time.

**Context**
Provides logic to reformat a list of Prometheus responses into the format expected by SUPReMM. Iterates over a response and at each timestamp yields a numpy array in the format that the preprocs and plugins expect. Preprocessors, range plugins, and firstlast plugins all require either different iterating or different output formats so the **Context** class also helps to facilitate this.

### src/supremm/datasource/prometheus/prommapping.py
This file contains two classes that help manage building Prometheus queries from the mapping configuration, querying Prometheus for existence of those metrics for a given job, and storing mapping information which is used in various steps of the summarization workflow.

### src/supremm/datasource/prometheus/promsummarize.py
The main logic for Prometheus summarization attempts to mimic the existing logic for PCP summarization. They are functionally different under the hood, but the overall workflow in this file should look very similar to **PCPSummarize**. The functional differences for Prometheus can be found in prominterface.py and prommapping.py. The high-level logic for Prometheus summarization is as follows:
For every node in a job:
--For every valid preproc/plugin:
----Translate PCP metric names -> Prometheus metric names
----If all required metrics are available:
------query Prometheus and load the entire response into memory
------Iterate over the Prometheus response and pass each yielded timestamp/data point to the preproc/plugins
----else the preproc/plugin fails

### src/supremm/preprocessors/ProcPrometheus.py
Inherits from the Proc.py preprocessor and only reimplements the process() method. Process data from Prometheus are formatted too differently and all reported processes from the CGroup exporter are constrained processes. This is the motivation for a separate preprocessor for Prometheus.

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1203222338177430

### CODE REVIEW

File: src/supremm/plugins/CpuUsage.py [hunk 1]
```diff
@@ -94,12 +94,29 @@ def computeallcpus(self):
     def computejobcpus(self):
         """ stats for the cores on the nodes that were assigend to the job (if available) """
 
-        proc = self._job.getdata('proc')
-
-        if proc == None:
+        # Check both proc and procprom preprocessors for CPU information
+        proc = None
+        for preproc in ('proc', 'procprom'):
```
Comment (jpwhite4) [inline src/supremm/plugins/CpuUsage.py:99]: The architectural design is that the plugins should be agnostic to the data source wherever possible. In this case does the CPU plugin really need to know whether the data came from prometheus or pcp? It appears not given that the code that uses the cpusallowed data is unchanged.

Please revert this change to the cpu plugin and just have the prometheus preprocessor use 'proc' as the identifier.

File: src/supremm/preprocessors/ProcPrometheus.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+from supremm.preprocessors.Proc import Proc
+from supremm.linuxhelpers import parsecpusallowed
+
+
+class ProcPrometheus(Proc):
+    """ Parse and analyse the proc information for a job. Supports parsing the cgroup information
+        from SLRUM and PBS/Torque (if available).
+    """
+
+    name = property(lambda x: "procprom")
```
Comment (jpwhite4) [inline src/supremm/preprocessors/ProcPrometheus.py:15]: should be 'proc' not 'procprom'

see comment in the cpuusage plugin

File: src/supremm/xdmodaccount.py [hunk 3]
```diff
@@ -1,4 +1,5 @@
 """ Implementation for account reader that gets data from the XDMoD datawarehouse """
+import sys #debug
```
Comment (jpwhite4) [inline src/supremm/xdmodaccount.py:2]: Please double check that you want to include this new line - it looks like it is left over from some debugging while developing the code.

File: src/supremm/datasource/factory.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+class DatasourceFactory():
+    """ Datasource class helper """
+
+    def __init__(self, preprocs, plugins, resconf):
+
+        if resconf["datasource"] == "pcp":
+            self._datasource = PCPDatasource(preprocs, plugins)
+        elif resconf["datasource"] == "prometheus":
+            self._datasource = PromDatasource(preprocs, plugins, resconf)
```
Comment (jpwhite4) [inline src/supremm/datasource/factory.py:13]: Perhaps throw an exception with a useful error message if the resconf['datasource'] is not a valid value.

File: src/supremm/datasource/pcp/pcpsummarize.py [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+                if result != None:
+                    ctx.pmFreeResult(result)
+
+        analytic.status = "complete"
+
+    def logerror(self, archive, analyticname, pmerrorcode):
+        """
+        Store the detail of archive processing errors
+        """
+        #if analyticname != "gpu":
```
Comment (jpwhite4) [inline src/supremm/datasource/pcp/pcpsummarize.py:261]: remove commented out code.

File: src/supremm/datasource/prometheus/promdatasource.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+            if (datetime.datetime.now() - job.end_datetime) > datetime.timedelta(seconds=force_timeout):
+                force_success = True
+
+        return s, jobmeta.mdata, success or force_success, jobmeta.error
+
+    def cleanup(self, opts, job):
+        # Nothing to be done for Prometheus
+        pass
+
+def parse_scrape_interval(interval):
```
Comment (jpwhite4) [inline src/supremm/datasource/prometheus/promdatasource.py:100]: This looks like an unused function. I think it should be removed.

File: src/supremm/datasource/prometheus/prominterface.py [hunk 7]
```diff
@@ -0,0 +1,476 @@
...
+    @staticmethod
+    def build_info(client, test_url):
+        """ Query server build info. Test connection to server. """
+
+        endpoint = "/api/v1/status/buildinfo"
+        url = urlparse.urljoin(test_url, endpoint)
+
+        r = client.get(url)
+        if r.status_code != 200:
+            print(str(r.content))
```
Comment (jpwhite4) [inline src/supremm/datasource/prometheus/prominterface.py:40]: Please use logging class rather than `print`

File: src/supremm/datasource/prometheus/prominterface.py [hunk 8]
```diff
@@ -0,0 +1,476 @@
...
+            'end': end,
+            'step': self._step
+        }
+
+        endpoint = "/api/v1/query_range"
+        url = urlparse.urljoin(self._url, endpoint)
+
+        r = self._client.get(url, params=params)
+        if r.status_code != 200:
+            print(r.content)
```
Comment (jpwhite4) [inline src/supremm/datasource/prometheus/prominterface.py:83]: Please use the logging class rather than `print`

File: src/supremm/datasource/prometheus/promsummarize.py [hunk 9]
```diff
@@ -0,0 +1,301 @@
...
+            result = next(results)
+        except StopIteration:
+            analytic.status = "failure"
+            return
+        except requests.RequestException as exp:
+            analytic.status = "failure"
+            raise exp
+        except Exception as exp:
+            analytic.status = "failure"
+            raise exp
```
Comment (aaronweeden) [inline src/supremm/datasource/prometheus/promsummarize.py:215]:
```suggestion
        except Exception as exp:
            analytic.status = "failure"
            raise exp
```
These `except` blocks can be combined since `requests.RequestException` inherits from `Exception`.

Comment (aaronweeden) [inline src/supremm/datasource/prometheus/promsummarize.py:215]: Duplicated code that could be moved to a method (see below).

File: src/supremm/datasource/prometheus/promsummarize.py [hunk 10]
```diff
@@ -0,0 +1,301 @@
...
```
Comment (aaronweeden) [inline src/supremm/datasource/prometheus/promsummarize.py:231]: Duplicated code that could be moved to a method (see above).

File: src/supremm/datasource/prometheus/promsummarize.py [hunk 11]
```diff
@@ -0,0 +1,301 @@
...
+                if False == self.runcallback(analytic, result, ctx, mdata):
+                    break
+            except StopIteration:
+                done = True
+            except requests.RequestException as exp:
+                analytic.status = "failure"
+                raise exp
+            except Exception as exp:
+                analytic.status = "failure"
+                raise exp
```
Comment (aaronweeden) [inline src/supremm/datasource/prometheus/promsummarize.py:266]: Duplicate code that could be moved to a method (see above).

File: .circleci/config.yml [hunk 12]
```diff
@@ -1,32 +1,64 @@
...
+      - run:
+          name: Build services
+          command: docker compose -f ./tests/ci/srv/services.yml build
+      - run:
+          name: Start services
+          command: docker compose -f ./tests/ci/srv/services.yml up -d
       - run:
           name: Run Bootstrap
-          command: ./tests/integration_tests/bootstrap.sh
+          command: ./tests/ci/test/bootstrap.sh $XDMOD_TEST_MODE
```
Comment (jpwhite4) [inline .circleci/config.yml:61]: The bootstrap.sh script does not appear to check its commandline arguments. Please double check this change

File: .circleci/config.yml [hunk 13]
```diff
@@ -73,4 +105,11 @@ workflows:
       - build:
           matrix:
             parameters:
-              install-type: ["fresh_install", "upgrade"]
+              build-type: ["rpm"]
+
+      - test:
+          matrix:
+            parameters:
+              install-type: ["rpm"]
```
Comment (jpwhite4) [inline .circleci/config.yml:113]: Why is the semantic meaning of install-type changed? It use to distingush between  fresh install and an upgrade, but now it looks like it is set to something completely different.

File: src/supremm/account.py [hunk 14]
```diff
@@ -178,8 +178,8 @@ class DbAcct(Accounting):
     Helper class to get job records from the store
     """
 
-    def __init__(self, resource_id, conf):
-        super(DbAcct, self).__init__(resource_id, conf)
+    def __init__(self, resconf, conf):
```
Comment (jpwhite4) [inline src/supremm/account.py:181]: Please can you explain why this changed? It looks like the only use of resconf is to pull in the resource_id (which is what was being used before).

File: src/supremm/datasource/pcp/pcpsummarize.py [hunk 15]
```diff
@@ -0,0 +1,365 @@
...
+import traceback
+from supremm.plugin import NodeMetadata
+from supremm.rangechange import RangeChange, DataCache
+from supremm.summarize import Summarize
+from supremm.datasource.pcp.pcpcinterface import pcpcinterface
+
+import numpy
+import copy
+
+VERSION = "2.0.0"
```
Comment (jpwhite4) [inline src/supremm/datasource/pcp/pcpsummarize.py:20]: This version should only change if the output schema changes. Please don't change this!

File: src/supremm/datasource/pcp/pcpsummarize.py [hunk 16]
```diff
@@ -0,0 +1,365 @@
...
+                    if analytic.mode == "all":
+                        output[analytic.name] = analytic.results()
+                    if analytic.mode == "timeseries":
+                        timeseries[analytic.name] = analytic.results()
+            for analytic in self.firstlast:
+                if analytic.status != "uninitialized":
+                    output[analytic.name] = analytic.results()
+
+        output['summarization'] = {
+            "version": VERSION,
```
Comment (jpwhite4) [inline src/supremm/datasource/pcp/pcpsummarize.py:102]: this could be why your ookami data is not showing.

File: src/supremm/datasource/prometheus/promsummarize.py [hunk 17]
```diff
@@ -0,0 +1,301 @@
...
+
+import requests
+import numpy as np
+
+from supremm.datasource.prometheus.prominterface import PromClient, Context
+from supremm.plugin import NodeMetadata
+from supremm.summarize import Summarize
+
+
+VERSION = "2.0.0"
```
Comment (jpwhite4) [inline src/supremm/datasource/prometheus/promsummarize.py:14]: Don't change the version

File: src/supremm/summarize.py [hunk 18]
```diff
@@ -1,54 +1,27 @@
...
-import traceback
-from supremm.plugin import NodeMetadata
-from supremm.rangechange import RangeChange, DataCache
-from supremm.pcpcinterface import pcpcinterface
-
-import numpy
-import copy
-
-VERSION = "1.0.6"
+VERSION = "2.0.0"
```
Comment (jpwhite4) [inline src/supremm/summarize.py:4]: Why is this constant (1) changed, and defined in at least 3 different places?

File: tests/integration_tests/supremm_setup_expect.py [hunk 19]
```diff
@@ -55,9 +69,9 @@ def main():
         p.expect("DB port")
         p.sendline()
         p.expect("DB Admin Username")
-        p.sendline()
+        p.sendline("xdmod")
         p.expect("DB Admin Password")
-        p.sendline()
+        p.sendline("xdmod123")
```
Comment (jpwhite4) [inline tests/integration_tests/supremm_setup_expect.py:74]: Please explain why this changed

File: src/supremm/datasource/pcp/pcpsummarize.py [hunk 20]
```diff
@@ -0,0 +1,365 @@
...
+
+        if len(timeseries) > 0:
+            timeseries['hosts'] = dict((str(idx), name) for name, idx, _ in self.job.nodearchives())
+            timeseries['version'] = TIMESERIES_VERSION
+            output['timeseries'] = timeseries
+
+        for preproc in self.preprocs:
+            result = preproc.results()
+            # Don't overrwrite proc data if ProcPrometheus is enabled
+            if preproc.name == 'procprom':
```
Comment (jpwhite4) [inline src/supremm/datasource/pcp/pcpsummarize.py:121]: Please see other comments about the pre-procs. We should not have prometheus-specfic code in the pcp code.

File: src/supremm/summarize.py [hunk 21]
```diff
@@ -1,54 +1,27 @@
...
-    nodeindex = property(lambda self: self._nodeidx)
-    archive = property(lambda self: self._archivedata)
-
-class Summarize(object):
-    """
-    Summarize class is responsible for iteracting with the pmapi python code
-    and managing the calls to the various analytics to process the data
+class Summarize(ABC):
+    """ Abstract base class describing the job summarization interface.
+        Currently only interfaces with PCP archives and is subject to change.
```
Comment (jpwhite4) [inline src/supremm/summarize.py:10]: Please double check this comment is correct. Why create an abstract base class with only one implementtaion?

File: src/supremm/account.py (hunk 14 above)
Comment (connersaeli) [reply src/supremm/account.py]: Resolved in https://github.com/ubccr/supremm/pull/259/commits/7a424e68517909a53032843e1673c6808030f269 and https://github.com/ubccr/supremm/pull/259/commits/9224350ea211d934aca42b8e496fce0162763b7c

File: tests/integration_tests/supremm_setup_expect.py (hunk 19 above)
Comment (connersaeli) [reply tests/integration_tests/supremm_setup_expect.py:74]: This change is here from the larger set of changes I was making to the tests that include an external Mariadb service (instead of included in the job_performance 10.0 image). This was the password for the service. Because that feature for the tests was reverted, I can just remove for now and add it back when the full testing changes are implemented.

File: src/supremm/summarize.py (hunk 18 above)
Comment (connersaeli) [reply src/supremm/summarize.py]: Resolved in:
https://github.com/ubccr/supremm/pull/259/commits/a34241f9050f4e6fcc5b9fb084199663ce3c9fe2
https://github.com/ubccr/supremm/pull/259/commits/e928967e850ddb22068e1e23ae33fa39e58e0b23
https://github.com/ubccr/supremm/pull/259/commits/c7b17840d6afddd9945104bdc48378a428543fb2

File: src/supremm/datasource/pcp/pcpsummarize.py (hunk 15 above)
Comment (connersaeli) [reply src/supremm/datasource/pcp/pcpsummarize.py]: Resolved in:
https://github.com/ubccr/supremm/pull/259/commits/a34241f9050f4e6fcc5b9fb084199663ce3c9fe2
https://github.com/ubccr/supremm/pull/259/commits/e928967e850ddb22068e1e23ae33fa39e58e0b23
https://github.com/ubccr/supremm/pull/259/commits/c7b17840d6afddd9945104bdc48378a428543fb2

File: src/supremm/datasource/pcp/pcpsummarize.py (hunk 16 above)
Comment (connersaeli) [reply src/supremm/datasource/pcp/pcpsummarize.py]: Resolved in:
https://github.com/ubccr/supremm/pull/259/commits/a34241f9050f4e6fcc5b9fb084199663ce3c9fe2
https://github.com/ubccr/supremm/pull/259/commits/e928967e850ddb22068e1e23ae33fa39e58e0b23
https://github.com/ubccr/supremm/pull/259/commits/c7b17840d6afddd9945104bdc48378a428543fb2

File: src/supremm/datasource/prometheus/promsummarize.py (hunk 17 above)
Comment (connersaeli) [reply src/supremm/datasource/prometheus/promsummarize.py]: Resolved in:
https://github.com/ubccr/supremm/pull/259/commits/a34241f9050f4e6fcc5b9fb084199663ce3c9fe2
https://github.com/ubccr/supremm/pull/259/commits/e928967e850ddb22068e1e23ae33fa39e58e0b23
https://github.com/ubccr/supremm/pull/259/commits/c7b17840d6afddd9945104bdc48378a428543fb2

File: src/supremm/summarize.py (hunk 21 above)
Comment (connersaeli) [reply src/supremm/summarize.py]: Resolved in https://github.com/ubccr/supremm/pull/259/commits/a34241f9050f4e6fcc5b9fb084199663ce3c9fe2

File: src/supremm/datasource/pcp/pcpsummarize.py (hunk 5 above)
Comment (connersaeli) [reply src/supremm/datasource/pcp/pcpsummarize.py:258]: Resolved in https://github.com/ubccr/supremm/pull/259/commits/a50f9f32a7e63d5804f3f01844149b6cc2936be9

File: src/supremm/xdmodaccount.py (hunk 3 above)
Comment (connersaeli) [reply src/supremm/xdmodaccount.py]: Resolved in https://github.com/ubccr/supremm/pull/259/commits/a50f9f32a7e63d5804f3f01844149b6cc2936be9

File: src/supremm/datasource/prometheus/promdatasource.py [hunk 22]
```diff
@@ -0,0 +1,121 @@
...
+        jobmeta = super().presummarize(job, conf, resconf, opts)
+
+        # Initialize client and test connection
+        if not self.client and not self.mapping:
+            self.client = PromClient(resconf)
+            if not self.client.connection:
+                jobmeta.result = 1
+                jobmeta.mdata["skipped_no_prom_connection"] = True
+                jobmeta.error = ProcessingError.PROMETHEUS_CONNECTION
+                logging.info("Skipping %s, skipped_no_prom_connection", job.job_id)
```
[the remark this answers is not in the document]
Comment (connersaeli) [reply src/supremm/datasource/prometheus/promdatasource.py:47]: Resolved in https://github.com/ubccr/supremm/pull/259/commits/15e080df0daf22aad8e2d9f34fe3cacd47ce39c5

File: src/supremm/datasource/prometheus/prominterface.py (hunk 8 above)
Comment (connersaeli) [reply src/supremm/datasource/prometheus/prominterface.py:83]: Resolved in https://github.com/ubccr/supremm/pull/259/commits/ba29736be72b3035bebd70ae7d0c636ac46ad87b

File: src/supremm/datasource/prometheus/prominterface.py (hunk 7 above)
Comment (connersaeli) [reply src/supremm/datasource/prometheus/prominterface.py:40]: Resolved in https://github.com/ubccr/supremm/pull/259/commits/ba29736be72b3035bebd70ae7d0c636ac46ad87b

File: src/supremm/datasource/prometheus/prominterface.py [hunk 23]
```diff
@@ -0,0 +1,476 @@
...
+            'query': query,
+            'time': time,
+        }
+
+        endpoint = "/api/v1/query"
+        url = urlparse.urljoin(self._url, endpoint)
+
+        r = self._client.get(url, params=params)
+        if r.status_code != 200:
+            print(str(r.content))
```
[the remark this answers is not in the document]
Comment (connersaeli) [reply src/supremm/datasource/prometheus/prominterface.py:63]: Resolved in https://github.com/ubccr/supremm/pull/259/commits/ba29736be72b3035bebd70ae7d0c636ac46ad87b

File: src/supremm/datasource/prometheus/promdatasource.py (hunk 6 above)
Comment (connersaeli) [reply src/supremm/datasource/prometheus/promdatasource.py:100]: Resolved in https://github.com/ubccr/supremm/pull/259/commits/ba29736be72b3035bebd70ae7d0c636ac46ad87b

File: tests/integration_tests/supremm_setup_expect.py (hunk 19 above)
Comment (connersaeli) [reply tests/integration_tests/supremm_setup_expect.py]: Resolved in https://github.com/ubccr/supremm/pull/259/commits/dcfe32158a538fbb3ec8cd10ad18dd3cea87835a

File: .circleci/config.yml (hunk 13 above)
Comment (connersaeli) [reply .circleci/config.yml]: Attempt to resolve in consecutive commits starting at https://github.com/ubccr/supremm/pull/259/commits/701e98cd093480c2693c8031cfd85cadc938d92c and ending at https://github.com/ubccr/supremm/pull/259/commits/2f4a8ea1af540891c311dc82746c16604b8e0718. Testing step now only does tests for a given installation parameter (rpm, wheel, or source).

File: src/supremm/datasource/pcp/pcpsummarize.py (hunk 20 above)
Comment (connersaeli) [reply src/supremm/datasource/pcp/pcpsummarize.py]: I have a fix for this in https://github.com/ubccr/supremm/pull/259/commits/7ac9f16c393045118491a9a9ccdadc678f1d148a. I removed these lines and successfully tested this change on Rocky 8 Ookami. The previous issue that the code above meant to resolve is now resolved by the new commit.

Comment (connersaeli) [reply src/supremm/datasource/pcp/pcpsummarize.py]: This was also tested with Prometheus data on supremm-dev in addition to the PCP data on Rocky 8 Ookami.

File: src/supremm/preprocessors/ProcPrometheus.py (hunk 2 above)
Comment (connersaeli) [reply src/supremm/preprocessors/ProcPrometheus.py]: I have a fix for this in this commit https://github.com/ubccr/supremm/commit/7ac9f16c393045118491a9a9ccdadc678f1d148a

File: src/supremm/plugins/CpuUsage.py (hunk 1 above)
Comment (connersaeli) [reply src/supremm/plugins/CpuUsage.py]: These changes have been reverted in https://github.com/ubccr/supremm/pull/259/commits/dd2ec42481bc4e9cd2bb5e5dc76427262ed4f8b2

File: src/supremm/datasource/factory.py (hunk 4 above)
Comment (connersaeli) [reply src/supremm/datasource/factory.py:15]: Resolved in https://github.com/ubccr/supremm/pull/259/commits/bcea3a659671902a42e250936536b72d96e6a352, https://github.com/ubccr/supremm/pull/259/commits/cfcd5ba2f7f6785e91664f0325fa17493056f74f, and https://github.com/ubccr/supremm/pull/259/commits/fc08e7a5784945738b8c4973213d153511b7c787

**Review by jpwhite4 — DISMISSED** [review]

File: config/config.json [hunk 24]
```diff
@@ -38,7 +38,8 @@
             "resource_id": 1,
             "batch_system": "XDMoD",
             "hostname_mode": "hostname",
-            "pcp_log_dir": "/data/pcp-logs/my_cluster_name",
+            "datasource": "pcp",
+	    "pcp_log_dir": "/data/pcp-logs/my_cluster_name",
```
Comment (jpwhite4) [inline config/config.json:42]: please use spaces rather than tabs in json files.

File: src/supremm/datasource/pcp/pcpdatasource.py [hunk 25]
```diff
@@ -0,0 +1,101 @@
...
+            else:
+                logging.error("Failure extracting logs for job %s", job.job_id)
+                return None
+
+        return jobmeta
+
+    def summarizejob(self, job, jobmeta, conf, opts):
+        preprocessors, analytics = super().summarizejob(job, jobmeta, conf, opts)
+
+        s = PCPSummarize(preprocessors, analytics, job, conf, opts["fail_fast"], PCP_STR)
```
Comment (jpwhite4) [inline src/supremm/datasource/pcp/pcpdatasource.py:60]: The prototype for PCPSumarize is this:
```
def __init__(self, preprocessors, analytics, job, config, datasource, fail_fast=False):
```

the function arguments don't match what is being called here: PCP_STR and fail_fast have been transposed.

My question is why do you pass in PCP_STR to PCPSummarize? Is there ever a scenario where the pcp-specific summarization code will use a different datasource string than PCP_STR?

**Review by jpwhite4 — CHANGES_REQUESTED** [review]
